### PR TITLE
Set matmul precision to medium if AMP is enabled for faster training

### PR DESCRIPTION
**Pull request by Y-T-G**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Sets a default mixed-precision matrix multiplication mode to "medium" whenever AMP is enabled during training for more stable and efficient computation ⚙️

### 📊 Key Changes
- 🧮 In `Trainer._setup_train`, when `self.amp` (Automatic Mixed Precision) is enabled, the code now calls:
  - `torch.set_float32_matmul_precision("medium")`
- ✅ AMP boolean handling and gradient scaler initialization remain unchanged; this PR only adjusts the global matmul precision setting under AMP.

### 🎯 Purpose & Impact
- ⚖️ Improves the balance between speed and numerical accuracy for float32 matrix multiplications when using AMP.
- 🧷 Can reduce potential numerical instability or precision-related quirks during training, especially on newer PyTorch versions and GPUs.
- 🚀 Offers more predictable and generally safer default behavior for users who enable AMP, with potential mild performance benefits and more stable training runs.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated review assistant, and a Ultralytics engineer will be with you shortly to provide further feedback.

To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/engine/trainer.py [hunk 1]
```diff
@@ -303,6 +303,8 @@ def _setup_train(self):
         if RANK > -1 and self.world_size > 1:  # DDP
             dist.broadcast(self.amp.int(), src=0)  # broadcast from rank 0 to all other ranks; gloo errors with boolean
         self.amp = bool(self.amp)  # as boolean
+        if self.amp:
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py:306]: 💡 **MEDIUM**: `torch.set_float32_matmul_precision` was introduced in later PyTorch releases, so calling it unconditionally whenever AMP is enabled will raise `AttributeError` on older, but still supported, torch versions. Consider guarding this call with `hasattr(torch, "set_float32_matmul_precision")` (and potentially catching runtime errors) to keep training from crashing on valid setups.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Please add a version/attribute guard around the new matmul-precision tweak so older supported torch installs don’t crash when AMP is enabled.

💬 Posted 1 inline comment

**Comment by Y-T-G** [comment]
Seems to just be the same speed on A100
